angular-app/firestore.rules

```diff
@@ -1,7 +1,25 @@
+rules_version = '2';
 service cloud.firestore {
   match /databases/{database}/documents {
-    match /{document=**} {
-      allow read, write;
+    function isSignedIn() {
+    	return request.auth.uid != null;
+    }
+    
+    match /projects/{projectID} {
+    	function isOwnedBy(userId) {
+    		return get(/databases/$(database)/documents/projects/$(projectID)).data.roles[userId] == "owner";
+    	}
+    	function hasCollaborator(userId) {
+    		return get(/databases/$(database)/documents/projects/$(projectID)).data.roles[userId] in ["owner", "editor"];
+    	}
+    
+    	allow create: if isSignedIn();
+    	allow read, update: if hasCollaborator(request.auth.uid);
+			allow delete: if isOwnedBy(request.auth.uid);
+      
+      match /files/{document=**} {
+      	allow read, write: if hasCollaborator(request.auth.uid);
```

Comment:
A little confused by this rule, though I admit I'm not completely familiar with the `files/` collection. Does this mean that if a user is a collaborator on any project, they have access to all files in the collection? Shouldn't this be scoped to projects they have access to?

Reply:
`files/` is a subcollection of a project; that is, each project has its own independent collection of files. Since this rule is nested under a particular project ID (and that specific projectID is used in the `hasCollaborator` function), I believe that files should only be accessible if you are a collaborator on that project.

Reply:
Thanks for the explanation.